junit-jupiter-api/src/main/java/org/junit/jupiter/api/Assertions.java

```diff
@@ -1479,12 +1560,102 @@ private static void failArraysNotEqual(Object expected, Object actual, Deque<Int
 		return result;
 	}
 
-	private static String formatIndexes(Deque<Integer> indexes) {
-		if (indexes == null || indexes.isEmpty()) {
-			return "";
+	// --- assertIterableEquals helpers ----------------------------------
+
+	private static void assertIterableEquals(Iterable<?> expected, Iterable<?> actual, Deque<Integer> indexes,
+			Supplier<String> messageSupplier) {
+
+		if (expected == actual) {
+			return;
 		}
-		String indexesString = indexes.stream().map(Object::toString).collect(joining("][", "[", "]"));
-		return " at index " + indexesString;
+		assertIterablesNotNull(expected, actual, indexes, messageSupplier);
+
+		Iterator<?> expectedIterator = expected.iterator();
+		Iterator<?> actualIterator = actual.iterator();
+
+		int processed;
+		for (processed = 0; expectedIterator.hasNext() && actualIterator.hasNext(); processed++) {
```

Comment:
Yes I could. I can see why a while loop would be easier to read/maintain in this case, so I'll make that change ASAP.